### PR TITLE
BUG: fix wrong masked median for some special cases

### PR DESCRIPTION
**Pull request by juliantaylor**
the masked nans which are equivalent to valid infs must be replaced
with infs earlier otherwise the inf is lost in the masked sum of the low
and high part.
Closes gh-8487

### CODE REVIEW

File: numpy/ma/extras.py [hunk 1]
```diff
@@ -758,6 +758,19 @@ def _median(a, axis=None, out=None, overwrite_input=False):
     ind[axis] = np.minimum(h, asorted.shape[axis] - 1)
     high = asorted[tuple(ind)]
 
+    def replace_masked(s):
+        # Replace masked entries with minimum_full_value unless it all values
```
Comment (charris) [inline numpy/ma/extras.py]: Needs normal docstring.

Comment (charris) [inline numpy/ma/extras.py]: Oh, never mind, is local function.

**Comment by charris** [comment]
This whole median thing is getting out of hand. I'm thinking we should have some underlying functions, median, var, mean, etc.,  that simply take an `array` and a `where` , or `mask` and which could be used for both nan* and ma functions. Note, *not* the current masked versions, they are too specialized.

**Comment by charris** [comment]
Thanks Juiian.

**Comment by juliantaylor** [comment]
that doesn't work as the number of masked values per dimension may vary, of course you could just do everything with a loop over 1d arrays like nanmedian does for large dimensions, but its incredibly slow for small ones.

**Comment by juliantaylor** [comment]
just writing the thing in C might be a more reasonable option.

**Comment by charris** [comment]
I don't see what would be the problem. Basically, all of these routines reduce to masked manipulations, the nanfunction just produce that mask using `isnan`. So if we have efficient lower level functions, independent of the current masked functions, to deal with that we should be good.

**Comment by charris** [comment]
Want to do the backport?

**Comment by juliantaylor** [comment]
that is what nanmedian does for small dimension, it calls masked median with a mask generated by isnan.
I can do the backport but not today.

**Comment by charris** [comment]
Effectively, so do all the other nan functions. The point is, there are common lower level operations that we should think about implementing.

**Pull request merged**